Simperium/SPAuthenticationValidator.m

```diff
@@ -7,45 +7,111 @@
+    }
+
+    return NO;
 }
 
-- (BOOL)validatePasswordSecurity:(NSString *)password {
-    if ([password length] < self.minimumPasswordLength) {
+- (BOOL)validatePasswordWithUsername:(NSString *)username password:(NSString *)password error:(NSError **)error {
+    if (password.length < self.strongMinimumPasswordLength) {
+        if (error) {
+            NSString *description = NSLocalizedString(@"Password must contain at least %d characters", comment: @"Message displayed when password is too short. Please preserve the Percent D!");
+            NSString *formattedDescription = [NSString stringWithFormat:description, self.strongMinimumPasswordLength];
+
+            *error = [NSError sp_errorWithDomain:SPAuthenticationErrorDomain
+                                            code:SPAuthenticationErrorsPasswordTooShort
+                                     description:formattedDescription];
+        }
+        return NO;
+    }
+
+    if ([username isEqualToString:password]) {
+        if (error) {
+            NSString *description = NSLocalizedString(@"Password cannot match email", @"Message displayed when password is invalid (Signup)");
+            *error = [NSError sp_errorWithDomain:SPAuthenticationErrorDomain
+                                            code:SPAuthenticationErrorsPasswordMatchesUsername
+                                     description:description];
+        }
+
+        return NO;
+    }
+
+
+    if ([password containsString:NSString.sp_newline] || [password containsString:NSString.sp_tab]) {
+        if (error) {
+            NSString *description = NSLocalizedString(@"Password must not contain tabs nor newlines", comment: @"Message displayed when a password contains a disallowed character");
+            *error = [NSError sp_errorWithDomain:SPAuthenticationErrorDomain
+                                            code:SPAuthenticationErrorsPasswordContainsInvalidCharacter
+                                     description:description];
+        }
+
         return NO;
     }
     
     // Could enforce other requirements here
     return YES;
 }
 
+- (BOOL)validatePasswordConfirmation:(NSString *)confirmation password:(NSString *)password error:(NSError **)error {
+    if ([confirmation isEqualToString:password]) {
+        return YES;
+    }
+
+    if (error) {
+        NSString *description = NSLocalizedString(@"Passwords do not match", @"Password Validation: Confirmation doesn't match");
+        *error = [NSError sp_errorWithDomain:SPAuthenticationErrorDomain
+                                        code:SPAuthenticationErrorsConfirmationDoesntMatch
+                                 description:description];
+    }
+
+    return NO;
+}
+
+- (BOOL)mustPerformPasswordResetWithUsername:(NSString *)username password:(NSString *)password {
+    BOOL isValidUsername = [self validateUsername:username error:nil];
+    BOOL isValidLegacyPassword = (password.length >= self.legacyMinimumPasswordLength);
```

Comment:
Do we need this check because legacy passwords might contain tabs or new lines?  Not sure there is any other need for it so I wanted to double check.

Reply:
Actually the logic behind this one is:

A. Is the password **now** considered unsafe?
B. Was the password **then** considered safe?
C. Is the password valid (does the backend let us thru?)

In that case, we'd want to enter the PW Reset flow. I'll add a comment to document this, thanks for asking!!

Reply:
Hmm... Well... the check for `isValidLegacyPassword` is just checking the length. So if the length is too short, then its also going to be too short when checking `isValidStrongPassword`, so if `isValidLegacyPassword` is false then `isValidStrongPassword` will be false, so in this scenario the check is unnecessary.  
On the other hand, if `isValidLegacyPassword` is true, `isValidStrongPassword` can still be false. 
I guess my confusion about this check is that there never seems to be a scenario where we get:
isValidUsername = true
isValidLegacyPassword = false
isValidStrongPassword = true
which would be why we'd want to have that check. 
What am I missing?

Reply:
OH!!!! You're absolutely right. That scenario is not possible, yet, `isValidStrongPassword` is calculated anyways (and will be false).

I'll prettify the checks (I was aiming a bit for compact), but you're 100% on point. 

Thank yoouu!!